File: ds/fenwickTree-2.py
class FenwickTree2:
    def __init__(self, arr):
        n = len(arr)
        self.arr = arr
        self.ft = [0]*(n)
    
    @staticmethod
    def LSOne(s):
        return s & (-s)

    def __update(self, i, v):
        while i < len(self.ft):
            self.ft[i] += v
            i += FenwickTree2.LSOne(i)

    def updateRange(self, l, r, v):
        self.__update(l, v)
        self.__update(r + 1, -v)

    def query(self, i):
        k = i
        sm = 0
        while i:
            sm += self.ft[i]
            i -= FenwickTree2.LSOne(i)
        return sm + self.arr[k]

if __name__ == "__main__":
    arr = [0] + [5, 9, 6, 1, 8, 4, 3, 7, 2, 8]
    ft = FenwickTree2(arr)
    print(ft.query(5))
    ft.updateRange(4, 6, 1)
    print(ft.query(5))

File: java/masterchief/src/main/java/ds/fenwickTree/FenwickTree2.java
package ds.fenwickTree;

/**
 * This is a Fenwick tree implementation
 * with Range Updates & Point Queries
 */
public class FenwickTree2 {
    private int[] arr;
    private int[] ft;

    public FenwickTree2(int[] arr){
        this.arr = new int[arr.length + 1];
        // insert extra 0 at the start of the base array
        for (int i = 1; i < arr.length; i++) {
            this.arr[i] = arr[i];
        }
        // ft arr is initialized with zeroes
        this.ft = new int[arr.length + 1];
    }

    public void update(int left, int right, int v) {
        this.update(left, v);
        this.update(right + 1, -v);
    }

    public void update(int p, int v) {
        for (; p <= this.ft.length; p += p & -p) {
            this.ft[p] += v;
        }
    }

    /**
     * Query for value at that position
     * 1-based indexing
     * @return
     */
    public int query(int p) {
        return this.arr[p] + this.queryFt(p);
    }

    private int queryFt(int p){
        int sum = 0;
        for (; p > 0; p -= p & -p) {
            sum += this.ft[p];
        }
        return sum;
    }

}
